pymeasure/adapters/protocol.py

```diff
@@ -30,16 +30,96 @@
 log.addHandler(logging.NullHandler())
 
 
+def to_bytes(command):
+    """Change `command` to a bytes object"""
+    if isinstance(command, (bytes, bytearray)):
+        return command
+    elif command is None:
+        return b""
+    elif isinstance(command, str):
+        return command.encode("utf-8")
+    elif isinstance(command, (list, tuple)):
+        try:
+            return bytes(command)
+        except TypeError:
+            raise
+    elif isinstance(command, (int, float)):
+        return str(command).encode("utf-8")
+    raise TypeError(f"Invalid input of type {type(command).__name__}.")
+
+
 class ProtocolAdapter(Adapter):
     """ Adapter class testing command exchange without instrument hardware.
 
+    :param list of lists comm_pairs: List of message pairs. First message is
+        the written one, the second one is the read message, it is optional.
+        'None' indicates that this message (write or read) does not exist.
     :param kwargs: TBD key-word arguments
     """
 
-    def __init__(self, preprocess_reply=None, **kwargs):
-        super().__init__(preprocess_reply=preprocess_reply)
+    def __init__(self, comm_pairs=[], preprocess_reply=None, **kwargs):
+        """Generate the adapter and initialize internal buffers."""
+        assert isinstance(comm_pairs, (list, tuple)), (
+            "Parameter comm_pairs has to be a list or tuple.")
+        super().__init__(preprocess_reply=preprocess_reply, **kwargs)
+        self._read_buffer = b""
+        self._write_buffer = b""
+        self.comm_pairs = comm_pairs
+        self._index = 0
         # TODO: Make this skeleton implementation workable
 
+    def write(self, command):
+        """Compare the command with the expected one and fill the buffer."""
+        pair = self.comm_pairs[self._index]
+        self._index += 1
+        assert to_bytes(pair[0]) == self._write_buffer + to_bytes(command), (
+                f"Command '{command}' written, but {pair[0]} expected.")
+        self._write_buffer = b""
+        try:
+            self._read_buffer = to_bytes(pair[1])
+        except IndexError:
+            # No response in the pair.
+            self._read_buffer = b""
+
+    def write_bytes(self, content):
+        """Write the bytes `content`. If a command is full, fill the read."""
+        self._write_buffer += content
```

Comment:
Is there a valid situation when we _append_ to the write buffer? Shouldn't whatever is in there already been consumed in a previous operation?

Reply:
Understood, thanks!